### ISSUE TITLE
Image support on Pushover 3.0

### ISSUE THREAD

**Issue by jonferreira** [qbit/node-pushover#20]
Hi,
Any plans to include image attachments support added on 3.0?

Should be a case of adding an image in multipart/form-data

**Comment by qbit** [comment]
Yeah the main goal is to stay in-line with the API - no ETA though.

**Comment by imbrianj** [comment]
I've been trying to get this to work without relying on an external library.  I have the multipart POST working, but have not been able to get the image to properly display.  It is being sent, message properly displayed and the image detected - but it renders as a placeholder.  You can see what I have here - it's a distillation of work in my MIT licensed project.

https://gist.github.com/imbrianj/af09af8cc06781e4f410062b38514c95

**Comment by jonferreira** [comment]
That does happen to me as well once in a while. Could be working perfectly for a week and then one day I just see the placeholder...

That said I have the feeling it's something on the app itself (android here). If you click on the placeholder and refresh the list do you see the image?

Also, I can send you my fully working nodes if you want

**Comment by imbrianj** [comment]
@jonferreira if you have something working, I'd love to give it a look over!

**Comment by jonferreira** [comment]
@imbrianj there you go https://pastebin.com/ttX7TiVL

**Comment by imbrianj** [comment]
@jonferreira huge help - thank you!  I wasn't setting up my Buffer correctly, so the image was getting lost.  I've updated my home automation system to now send me pics.

You can see what I've got here: https://github.com/imbrianj/switchBoard/blob/master/devices/pushover/controller.js

**Comment by jonferreira** [comment]
glad I could help

**Comment by SebastianSchultz** [comment]
I woud be glad if you can implement the sending of images :)

**Comment by Kev0123** [comment]
Any news on this? Would be a big step forward!

**Comment by damoclark** [comment]
Hi @Kev0123 / All,

I just submitted a PR implementing this functionality.  See #23 #

Damo.

**Comment by qbit** [comment]
Can everyone give #24 a spin - It implements attachments without external dependencies.

Images can be attached to the object passed to `send()` like so:
```
fs.readFile('test/test_img.png', function(err, data) {
  var o = {
    name: 'pushover.png',
    data: data
  }
  var msg = {
    message: 'test from ' + process.argv[1],
    sound: 'magic',
    title: 'Image loaded async',
    file: o
  }
  p.send(msg, function (err, result, res) {
    console.log('error', err)
    console.log('result', result)
    console.log('res.headers', res.headers)
    // process.exit(0);
  })
})
```

OR

```
var msg = {
  message: 'test from ' + process.argv[1],
  sound: 'magic',
  title: 'Image loaded sync',
  file: 'test/test_img.png'
}
p.send(msg, function (err, result, res) {
  console.log('error', err)
  console.log('result', result)
  console.log('res.headers', res.headers)
  // process.exit(0);
})
```

**Comment by damoclark** [comment]
Hey Aaron,

A few of quick things with #24 

1. There is a bug with the device option specifically when attaching an image using a `Buffer`. When you specify a device, the message still goes to all devices.

Here is how to reproduce it (substituting your own device name):

```js
var p = new Push({
  user: process.env['PUSHOVER_USER'],
  token: process.env['PUSHOVER_TOKEN'],
  update_sounds: false,
  device: 'devicename',
  debug: true
})

fs.readFile('test_img.png', function(err, data) {
  var o = {
    name: 'test_img.png',
    data: data
  }

  var msg = {
    message: 'test from ' + process.argv[1],
    sound: 'magic',
    title: 'Image loaded async',
    file: o
  }

  p.send(msg, function (err, result, res) {
    console.log('error', err)
    console.log('result', result)
    console.log('res.headers', res.headers)
  })
})
```


2. In some situations (i.e. Buffer-stored images), I don’t have a filename.  Omitting that property results in the library sending the header `filename="undefined"` which the pushover API server seems to happily accept.  I don’t think it pays attention to the filename of the upload but one is required as per the RFC.  

3. Pushover also seems to happily accept `application/octet-stream` as the content type as I don’t always know what the mime type of the images will be (e.g. jpg or png etc).  I’m pretty sure pushover validates everything that is sent to them anyway. ;) 

Hope this feedback is helpful.

Cheers,
Damo.

**Comment by qbit** [comment]
> Incorrect paths for the image files in the test suite

The paths are relative because the assumption is npm is doing the tests from the root of the repo. ( https://travis-ci.org/qbit/node-pushover )

> There is also a deprecation warning when using node.js v10.12.0. 

Updated this to use `.from`

> There is a bug with the device option specifically when attaching an image using a Buffer. When you specify a device, the message still goes to all devices.

I do believe this was because of async. I have updated the examples to not load `p` outside of the `fs.readFile` call.
